File: src/main/kotlin/io/github/hxxniverse/hobeaktown/feature/keycard/commands/KeycardCommand.java
package io.github.hxxniverse.hobeaktown.feature.keycard.commands;

import io.github.hxxniverse.hobeaktown.feature.keycard.connection.DatabaseManager;
import net.kyori.adventure.text.Component;
import org.bukkit.Material;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.jetbrains.annotations.NotNull;

import java.sql.SQLException;

public class KeycardCommand implements CommandExecutor {
    private final DatabaseManager databaseManager;

    public KeycardCommand(DatabaseManager databaseManager) {
        this.databaseManager = databaseManager;
    }

    @Override
    public boolean onCommand(@NotNull CommandSender commandSender, @NotNull Command command, @NotNull String s, @NotNull String[] strings) {
        if (!(commandSender instanceof Player player)) {
            commandSender.sendMessage("플레이어만 사용 가능한 명령어 입니다.");
            return true;
        }

        if (strings.length == 0) {
            player.sendMessage("키카드 관련 명령어: /키카드 [생성|철문|등록|아이템설정]");
            return true;
        }

        String subCommand = strings[0];

        switch (subCommand) {
            case "생성":
                if (strings.length != 3) {
                    player.sendMessage("키카드 명령어: /키카드 생성 <이름> <태그>");
                    return true;
                }
                String name = strings[1];
                String tag = strings[2];
                try {
                    if(databaseManager.isExistsKeyCard(name, tag)) player.sendMessage("이미 등록되어있는 키카드입니다.");
                    else {
                        databaseManager.insertKeyCard(name, tag);
                        player.sendMessage("키카드 " + name + " (태그: " + tag + ") 생성됨.");
                    }
                } catch (SQLException e) {
                    player.sendMessage(e.getMessage());
                }
                break;

            case "철문":
                if (strings.length != 2) {
                    player.sendMessage("키카드 명령어: /키카드 철문 <이름>");
                    return true;
                }
                String permission = strings[1];

                ItemStack itemInHand = player.getInventory().getItemInMainHand();
                if(itemInHand.getType() != Material.IRON_DOOR) {
                    player.sendMessage("철문을 손에 들고 명령어를 입력해 주세요.");
                    return true;
                }
                ItemMeta meta = itemInHand.getItemMeta();
                if (meta != null) {
                    meta.displayName(Component.text(permission + " 문"));
                    meta.addEnchant(Enchantment.LUCK, 1, true);
                    meta.addItemFlags(ItemFlag.HIDE_ENCHANTS);
                }
                NBTItem nbtItem = new NBTItem(itemInHand);
                nbtItem.getItem().setItemMeta(meta);
                nbtItem.setString("PermissionRegister", "true");
                nbtItem.setString("Permission", permission);
                nbtItem.applyNBT(itemInHand);

                player.sendMessage("철문에 " + permission + " 권한 등록됨.");
                break;

            case "등록":
                if (strings.length != 2) {
                    player.sendMessage("키카드 명령어: /키카드 등록 <이름>");
                    return true;
                }
                String permission2 = strings[1];

                ItemStack itemInHand2 = player.getInventory().getItemInMainHand();
                if(itemInHand2.isEmpty()) {
                    player.sendMessage("손에 물건을 들고 명령어를 입력해 주세요.");
                    return true;
                }

                ItemMeta meta2 = itemInHand2.getItemMeta();
                if (meta2 != null) {
                    meta2.displayName(Component.text(permission2 + " 키카드"));
                    meta2.addEnchant(Enchantment.LUCK, 1, true);
                    meta2.addItemFlags(ItemFlag.HIDE_ENCHANTS);
                }

                NBTItem nbtItem2 = new NBTItem(itemInHand2);
                nbtItem2.getItem().setItemMeta(meta2);
                nbtItem2.setString("PermissionRegister", "true");
                nbtItem2.setString("Permission", permission2);
                nbtItem2.applyNBT(itemInHand2);
                player.sendMessage("손에 들고 있는 아이템이 " + permission2 + " 키카드로 등록됨.");
                break;
        }

        return true;
    }
}

File: src/main/kotlin/io/github/hxxniverse/hobeaktown/feature/keycard/commands/KeyCardCommand.kt
package io.github.hxxniverse.hobeaktown.feature.keycard.commands

import io.github.hxxniverse.hobeaktown.feature.keycard.entity.KeyCard
import io.github.hxxniverse.hobeaktown.feature.user.Job
import io.github.hxxniverse.hobeaktown.feature.user.user
import io.github.hxxniverse.hobeaktown.util.ItemStackBuilder
import io.github.hxxniverse.hobeaktown.util.base.BaseCommand
import io.github.hxxniverse.hobeaktown.util.database.loggedTransaction
import io.github.hxxniverse.hobeaktown.util.extension.component
import io.github.hxxniverse.hobeaktown.util.extension.send
import io.github.monun.kommand.StringType.GREEDY_PHRASE
import io.github.monun.kommand.getValue
import io.github.monun.kommand.kommand
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import org.jetbrains.exposed.sql.transactions.transaction
import java.sql.SQLException
import java.util.EnumSet

class KeyCardCommand : BaseCommand {
    override fun register(plugin: JavaPlugin) {
        plugin.kommand {
            register("키카드") {
                requires { sender is Player }
                executes { player.sendMessage("명령어 도우미: /키카드 help") }
                then("help") {
                    executes {
                        """
                        |키카드 명령어
                        |  /키카드 생성 [태그] [이름]: [이름]에 [태그]를 등록하는 키카드를 생성합니다. (아이템 생성 X)
                        |  /키카드 철문 [이름]: 철문을 들고 해당 철문에 권한을 등록합니다. 설치 시 해당되는 키카드로 열고 닫을 수 있습니다.
                        |  /키카드 등록 [이름]: 손에 들고 있는 아이템이 해당 권한의 키카드가 됩니다.
                        """.trimIndent().also {
                            component(it).send(player)
                        }
                    }
                }
                then("생성") {
                    executes { player.sendMessage("명령어 사용법: /키카드 help"); }
                    then("job" to dynamicByEnum(EnumSet.allOf(Job::class.java))) {
                        then("name" to string(GREEDY_PHRASE)) {
                            executes {
                                val job: Job by it
                                val name: String by it

                                loggedTransaction {
                                    try {
                                        if (KeyCard.isExistsKeyCard(name, job)) {
                                            player.sendMessage("이미 등록되어있는 키카드입니다.")
                                        } else {
                                            KeyCard.new {
                                                this.name = name
                                                this.job = job
                                            }
                                            player.sendMessage("키카드 이름: $name 역할: ${job.name} 생성됨.")
                                        }
                                    } catch (e: SQLException) {
                                        player.sendMessage(e.message!!)
                                    }
                                }
                            }
                        }
                    }
                }
                then("철문") {
                    executes { player.sendMessage("명령어 사용법: /키카드 help"); }
                    then("name" to string(GREEDY_PHRASE)) {
                        executes {
                            val name: String by it

                            loggedTransaction {
                                val itemInHand = player.inventory.itemInMainHand
                                if (itemInHand.type != Material.IRON_DOOR) {
                                    player.sendMessage("철문을 손에 들고 명령어를 입력해 주세요.")
                                    return@loggedTransaction
                                }

                                if (KeyCard.isExistsKeyName(name)) {
                                    val keyCardDoorItem = ItemStackBuilder()
                                        .setType(itemInHand.type)
                                        .setDisplayName(
                                            component(name, NamedTextColor.BLUE)
                                                .append(component(" 철문", NamedTextColor.WHITE))
                                        )
                                        .addPersistentData("NameRegister", "true")
                                        .addPersistentData("Name", name)
                                        .build()

                                    player.inventory.setItemInMainHand(keyCardDoorItem)
                                    player.sendMessage("철문에 $name 이름 등록됨.")
                                } else player.sendMessage("'$name'은 등록되지 않은 이름입니다.")
                            }
                        }
                    }
                }
                then("등록") {
                    executes { player.sendMessage("명령어 사용법: /키카드 help"); }
                    then("name" to string(GREEDY_PHRASE)) {
                        executes {
                            val name: String by it

                            loggedTransaction {
                                var itemInHand = player.inventory.itemInMainHand
                                if (itemInHand.type == Material.IRON_DOOR) return@loggedTransaction
                                if (itemInHand.isEmpty) {
                                    player.sendMessage("아이템을 손에 들고 명령어를 입력해 주세요.")
                                    return@loggedTransaction
                                }

                                if (KeyCard.isExistsKeyName(name)) {
                                    itemInHand = ItemStackBuilder()
                                        .setType(itemInHand.type)
                                        .setDisplayName(
                                            component(name, NamedTextColor.BLUE)
                                                .append(component(" 키카드", NamedTextColor.WHITE))
                                        )
                                        .addPersistentData("NameRegister", "true")
                                        .addPersistentData("Name", name)
                                        .build()

                                    player.inventory.setItemInMainHand(itemInHand)
                                    player.sendMessage("아이템이 $name 키카드로 등록되었습니다.")
                                } else player.sendMessage("'$name'은 등록되지 않은 이름입니다.")
                            }
                        }
                    }
                }
                then("변경"){
                    requires { sender.isOp }
                    then("player" to player()){
                        then("job" to string()){
                            executes {
                                val player: Player by it
                                val job: String by it
                                loggedTransaction {
                                    player.user.job = Job.valueOf(job)
                                    player.sendMessage(component(job, NamedTextColor.BLUE).append(component("직업으로 변경되었습니다.", NamedTextColor.WHITE)))
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}